### PR TITLE
Fix main CI docs and minimum Pro selection

### PR DESCRIPTION
**Pull request by justin808**
## Why

`main` is currently red for two reproducible, unrelated reasons:

- `Validate llms-full.txt` on push run `28278196968` reports `llms-full-pro.txt is stale. Run node script/generate-llms-full.mjs and commit the result.` The triggering main commit updated Pro streaming docs but only regenerated `llms-full.txt`.
- `JS unit tests for Renderer package` on push run `28278196933` fails the Node 20 minimum-dependency Pro lane. That lane runs `script/convert` and then `pnpm --filter react-on-rails-pro test`; the converted `test:non-rsc` selector allowed the new `wrapServerComponentRenderer.client.performanceMarks.test.jsx` RSC/React 19 client-wrapper test into the React 18 profile, where it fails with `React.use is not defined`.

There was also a prevention gap in the docs-only main guard: when a docs-only push follows a previous main commit that failed only at the guard step, the guard reported the immediate parent as failed instead of walking back to the first parent with real CI state. That made the earlier main failure chain noisier and less actionable.

## What Changed

- Regenerated `llms-full-pro.txt` with `node script/generate-llms-full.mjs`.
- Broadened the `script/convert` minimum-dependency ignore pattern from the single `wrapServerComponentRenderer.client.test` file to the `wrapServerComponentRenderer.client` family, matching the existing intent to keep RSC client-wrapper tests out of the React 18 non-RSC Pro lane.
- Extracted the docs-only safety helper from the composite action into `check-previous-main.cjs`, added focused Node tests, and wired those tests into the lint workflow.
- The helper now follows first parents through guard-only docs failures, then allows/blocks based on the first commit with real CI state.

## Verification

- `node script/generate-llms-full.mjs --check`
- `node script/generate-llms-full.mjs --validate`
- `ruby -c script/convert`
- Temporary converted-install reproduction: `script/convert && pnpm install --no-frozen-lockfile && pnpm build && pnpm --filter react-on-rails-pro test`
  - Note: local reproduction used Node `v22.12.0`; hosted CI will confirm the exact Node 20 runner profile.
- Verified the converted `test:non-rsc` pattern excludes `wrapServerComponentRenderer.client.test`, `.performanceMarks`, `.imports`, and `.manifest` variants.
- `git diff --check`
- `node .github/actions/ensure-main-docs-safety/check-previous-main.test.cjs`
- `pnpm exec eslint .github/actions/ensure-main-docs-safety/check-previous-main.cjs .github/actions/ensure-main-docs-safety/check-previous-main.test.cjs`
- `pnpm exec prettier --check .github/actions/ensure-main-docs-safety/action.yml .github/actions/ensure-main-docs-safety/check-previous-main.cjs .github/actions/ensure-main-docs-safety/check-previous-main.test.cjs .github/workflows/lint-js-and-ruby.yml`
- `actionlint .github/workflows/*.yml`
- `script/ci-changes-detector origin/main`
- `pnpm start format.listDifferent`
- `pnpm run lint`
- `(cd react_on_rails && BUNDLE_GEMFILE=../Gemfile RUBOCOP_CACHE_ROOT=/tmp/rubocop-cache-main-ci bundle exec rubocop)`
- `codex review --base origin/main` reported no actionable regressions.
- Pre-push hooks passed.

`yamllint` is not installed in this local environment, so that specific optional YAML check was not run locally.

## Confidence note

- Validated: generated docs artifact check/validate, converted Pro minimum-dependency reproduction, action helper tests, JS lint, Prettier check, workflow lint, CI detector, full JS lint, OSS RuboCop, independent Codex review, git hooks.
- Evidence: local command output from the checks listed above; main failure evidence from push runs `28278196968` and `28278196933`.
- UNKNOWN: exact hosted Node 20 PR run result until GitHub Actions completes on this PR; whether the still-pending main `Rspec test for gem` run materializes a separate failure unrelated to this patch.
- Residual risk: low-to-moderate, mostly in the GitHub API path of the guard helper; covered by focused unit tests but still requires post-merge exercise with real Actions data.

## Summary by CodeRabbit

* **Documentation**
  * Added a new section on browser-visible performance marks for streamed React Server Components, including recorded mark types and key caveats.

* **Bug Fixes**
  * Improved docs-only CI skip safety by more precisely determining when prior checks contain real failures versus guard-only outcomes.

* **Chores**
  * Added and expanded automated tests covering the docs-safety decision logic, including pagination and parent-commit “look-through” scenarios.
  * Updated the lint workflow to run the new validation step for the action helper.

### CODE REVIEW

**Comment by justin808** [comment]
Post-merge Actions exercise tracker: #4230. This is required because the PR changes semantic GitHub Actions/composite-action behavior; it records the real Actions-data exercise expected after merge.

**Comment by justin808** [comment]
+ci-status

**Comment by justin808** [comment]
+ci-run-hosted

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+  const completedRunResults = await Promise.all(
+    completedRuns.map(async (run) => {
+      const {
+        data: { jobs },
+      } = await github.rest.actions.listJobsForWorkflowRun({
+        owner: context.repo.owner,
+        repo: context.repo.repo,
+        run_id: run.id,
+        per_page: 100,
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs]: Fixed in b0e0b5fa4. The helper now paginates listJobsForWorkflowRun via github.paginate.iterator, and testPaginatedJobsAreChecked covers a failed latest-attempt job on a later page.

File: .github/actions/ensure-main-docs-safety/action.yml [hunk 2]
```diff
@@ -20,186 +20,15 @@ runs:
...
-          );
+          const { checkPreviousMainCommitStatus } = require(`${process.env.ACTION_PATH}/check-previous-main.cjs`);
+
+          await checkPreviousMainCommitStatus({
+            github,
+            context,
+            core,
+            previousSha: process.env.PREVIOUS_SHA,
+            excludeWorkflowsInput: process.env.EXCLUDE_WORKFLOWS
+          });
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/action.yml:36]: Fixed in 0d6dea2e3. The composite action script now wraps the destructured require while preserving ACTION_PATH-based resolution.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+  return (
+    failed.length > 0 &&
+    failed.every(
+      (job) =>
+        job.name === 'detect-changes' &&
+        job.steps?.some(
+          (step) => step.name === 'Guard docs-only main pushes' && FAILURE_CONCLUSIONS.has(step.conclusion),
+        ),
+    )
+  );
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs:55]: Fixed in 0d6dea2e3. isGuardOnlyFailure now checks the failed steps in each failed detect-changes job and rejects mixed guard plus detector/setup failures; the mixedDetectChangesFailureJob regression test covers it.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+  const excludeWorkflows = parseExcludeWorkflows(excludeWorkflowsInput);
+
+  if (excludeWorkflows.length > 0) {
+    core.info(`Excluding workflows from failure checks: ${excludeWorkflows.join(', ')}`);
+  }
+
+  const guardOnlyTrail = [];
+
+  async function checkSha(shaToCheck, remainingGuardOnlyHops) {
+    if (remainingGuardOnlyHops < 0) {
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs]: Fixed in b0e0b5fa4. The hop guard now trips at <= 0, and testGuardOnlyHopLimitStopsAtConfiguredLimit covers the configured limit behavior.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    const pageRuns = response.data;
+    const relevantInPage = pageRuns.filter((run) => run.head_sha === sha);
+
+    if (relevantInPage.length > 0) {
+      workflowRuns.push(...relevantInPage);
+    }
+
+    // Early exit: if we found relevant runs and now see different SHAs,
+    // we've likely collected all runs for this commit.
+    if (workflowRuns.length > 0 && relevantInPage.length === 0) {
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs]: Fixed in b0e0b5fa4. listWorkflowRunsForSha no longer exits on the first non-matching page; testNonContiguousWorkflowRunPagesAreChecked covers target-SHA runs separated by other commits' runs.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+function isGuardOnlyFailure(jobs) {
+  const failed = failedJobs(jobs);
+
+  return (
+    failed.length > 0 &&
+    failed.every(
+      (job) =>
+        job.name === 'detect-changes' &&
+        job.steps?.some(
+          (step) => step.name === 'Guard docs-only main pushes' && FAILURE_CONCLUSIONS.has(step.conclusion),
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs]: Fixed in b0e0b5fa4. The guard job and step names are now module-level constants, and the tests use the same constants for the guard-only fixtures.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+
+function isGuardOnlyFailure(jobs) {
+  const failed = failedJobs(jobs);
+
+  return (
+    failed.length > 0 &&
+    failed.every(
+      (job) =>
+        job.name === 'detect-changes' &&
+        job.steps?.some(
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs]: Fixed in b0e0b5fa4. The classifier now requires an array of step data, treats missing steps conservatively as a real failure, and warnForMissingGuardSteps emits a diagnostic warning for failed guard jobs without step details.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+    jobs.push(...response.data.jobs);
+  }
+
+  return jobs;
+}
+
+function warnForMissingGuardSteps({ core, run, jobs }) {
+  for (const job of failedJobs(jobs)) {
+    if (job.name === GUARD_JOB_NAME && !Array.isArray(job.steps)) {
+      core.warning(
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs:99]: [auto-deferred]
Readability-only Map iteration cleanup deferred to avoid another optional-only CI/review cycle; current iteration/deletion behavior is spec-safe and already passes lint/tests.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+    repo: context.repo.repo,
+    ref: sha,
+  });
+
+  return response.data.parents?.[0]?.sha;
+}
+
+async function checkPreviousMainCommitStatus({
+  github,
+  context,
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs:182]: Already fixed in current head b0e0b5fa4: the guard condition is now <= 0, and testGuardOnlyHopLimitStopsAtConfiguredLimit covers the configured limit.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+async function evaluateCommitRuns({ github, context, core, sha, createdAfter, excludeWorkflows }) {
+  const workflowRuns = await listWorkflowRunsForSha({ github, context, sha, createdAfter });
+
+  if (workflowRuns.length === 0) {
+    return { status: 'no-runs', workflowRuns };
+  }
+
+  const latestByWorkflow = latestRunsByWorkflow(workflowRuns);
+
+  for (const [workflowId, run] of latestByWorkflow) {
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs:115]: Already fixed in current head b0e0b5fa4: listJobsForRun paginates listJobsForWorkflowRun with github.paginate.iterator, and testPaginatedJobsAreChecked covers a later-page failure.

File: .github/actions/ensure-main-docs-safety/check-previous-main.test.cjs [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+    pages: [[guardRun]],
+    jobsByRunId: {
+      13: [guardOnlyJob()],
+    },
+    parentsBySha: {
+      [previous]: parent,
+    },
+  });
+  const core = makeCore();
+
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.test.cjs:193]: Already fixed in current head b0e0b5fa4: testGuardOnlyHopLimitStopsAtConfiguredLimit exercises the limit-exceeded branch and asserts the diagnostic message.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+        [
+          `Cannot skip CI for docs-only commit because main commit ${shaToCheck} still has failing workflows:`,
+          details,
+          guardTrailDetails,
+          '',
+          'Fix these failures before pushing docs-only changes, or push non-docs changes to trigger hosted CI.',
+        ]
+          .filter(Boolean)
+          .join('\n'),
+      );
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs:257]: [auto-deferred]
Diagnostic blank-line polish deferred to avoid another optional-only CI/review cycle; the message still includes the failing workflows, optional guard trail, and remediation text.

File: .github/actions/ensure-main-docs-safety/check-previous-main.cjs [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+  }
+
+  const latestByWorkflow = latestRunsByWorkflow(workflowRuns);
+
+  for (const [workflowId, run] of latestByWorkflow) {
+    if (excludeWorkflows.includes(run.name)) {
+      core.info(`Excluding workflow "${run.name}" from failure checks (not a CI quality gate).`);
+      latestByWorkflow.delete(workflowId);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/actions/ensure-main-docs-safety/check-previous-main.cjs:120]: [auto-deferred]
Duplicate readability-only Map iteration cleanup deferred to avoid another optional-only CI/review cycle; current iteration/deletion behavior is spec-safe and already passes lint/tests.

**Comment by justin808** [comment]
## Address-review summary

Scan scope: full PR history through the review pass after commit `b0e0b5fa4`.

### Mattered
- CodeRabbit guard-only classifier finding: fixed in `0d6dea2e3`; mixed guard plus detector/setup failures now classify as real failures, with regression coverage.
- CodeRabbit action line-length finding: fixed in `0d6dea2e3`; composite action import is wrapped while preserving `ACTION_PATH` resolution.
- Greptile paginated jobs finding: fixed in `b0e0b5fa4`; workflow-run jobs are paginated and a later-page failing job is covered by test.
- Claude workflow-run pagination finding: fixed in `b0e0b5fa4`; the SHA scan no longer exits on the first non-matching page and has non-contiguous page coverage.
- Claude hop-limit finding: fixed in `b0e0b5fa4`; the guard trips at the configured limit and has regression coverage.
- Claude missing-step diagnostics: fixed in `b0e0b5fa4`; missing step data is treated conservatively and emits a warning.

### Optional
- Claude Map-iteration readability nits: auto-deferred in-thread as behavior-preserving polish to avoid an optional-only CI/review restart.
- Claude diagnostic blank-line polish: auto-deferred in-thread; the diagnostic still includes the failure details and remediation text.
- CodeRabbit strict test-mock nit: deferred as non-blocking test-harness polish; current regression tests cover the relevant API paths.

### Skipped
- Stale duplicate Claude comments about the old `< 0` condition, non-paginated jobs, and missing hop-limit coverage were replied to with the current-head fixes and resolved.

Deferred-work tracking: not tracking optional polish separately.

Next default scan starts after this comment. Say `check all reviews` to rescan the full PR.

**Comment by justin808** [comment]
Confidence note:
- Validated: local focused helper tests, docs artifact checks, JS lint/format/actionlint/RuboCop, temp converted Pro minimum-dependency reproduction, hosted optimized CI on PR head `b0e0b5fa4`, CodeQL, Claude review, CodeRabbit status, and all review threads resolved.
- Evidence: hosted checks on PR #4229 are green, including `Validate llms-full.txt`, `JS unit tests for Renderer package / build (22)`, `Lint JS and Ruby`, `Rspec test for gem`, Pro package/integration checks, generator/integration/playwright paths, actionlint, CodeQL, CodeRabbit, and Claude review.
- UNKNOWN: post-merge real Actions-data exercise for the docs-only guard; tracked in #4230 as required by repo policy for semantic GitHub Actions changes.
- Residual risk: low. The two current main failures are directly covered by hosted CI, and guard-helper edge cases raised in review now have regression tests.

Merge rationale: PR #4229 fixes the current red main checks (`llms-full-pro.txt` staleness and React 18 minimum-lane Pro test selection), plus hardens the docs-only guard that obscured the underlying failure chain. Release tracker #3823 is in `development` mode, the PR is cleanly mergeable, and no maintainer decision remains.
